### PR TITLE
[Examples] Finetune Falcon 7B and 40B Example

### PR DESCRIPTION
**Pull request by xzrderek**
I've added an example on how to fine-tune Falcon-7B-sharded (for those with less powerful GPUs), Falcon-7B, and Falcon-40B using SkyPilot and wrote a `README.md` (heavily borrowed from the Vicuna README).

Ideally, the repo referenced in my code, `https://github.com/xzrderek/sky-falcon.git`, should be a repo under skypilot-org, similar to the sky-llama example, where their code is stored [here](https://github.com/skypilot-org/sky-llama).

### CODE REVIEW

File: llm/falcon/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+## Prerequisites
+Install the latest SkyPilot and check your setup of the cloud credentials:
+```bash
+pip install git+https://github.com/skypilot-org/skypilot.git
+sky check
+```
+See the Falcon SkyPilot YAML for [training](train.yaml). Serving is currently a work in progress and a YAML will be provided for that soon! We are also working on adding an evaluation step to evaluate the model you finetuned compared to the base model.
+
+## Finetuning Falcon with SkyPilot
+Finetuning Falcon-7B and Falcon-40B require GPUs with 80GB memory, but Falcon-7b-sharded requires only 40GB memory. In other words, if your GPU does not have over 40 GB memory, i.e. `accelerators: A100:1`, we suggest using `ybelkada/falcon-7b-sharded-bf16`. `tiiuae/falcon-7b` and `tiiuae/falcon-40b` will need 80 GB memory. In our examples, we use A100-80GB and A100.
```
Comment (romilbhardwaj) [inline llm/falcon/README.md:19]: I tried fine-tuning `ybelkada/falcon-7b-sharded-bf16` on A100-40GB. The training seemed to result in a very spiky loss. Is this expected:

image

File: llm/falcon/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+```
+
+Currently, such `A100-80GB:1` spot instances are only available on AWS and GCP.
+
+[Optional] **To use on-demand `A100-80GB:1` instances**, which are currently available on Lambda Cloud, Azure, and GCP:
+```bash
+sky launch -c falcon -s train.yaml --no-use-spot
+```
+
+
```
Comment (romilbhardwaj) [inline llm/falcon/README.md:48]: Can we include here:
1. some screenshots from wandb of loss over time/epochs
2. How much time it takes
3. Approximate cost of fine-tuning

File: llm/falcon/README.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+See the Falcon SkyPilot YAML for [training](train.yaml). Serving is currently a work in progress and a YAML will be provided for that soon! We are also working on adding an evaluation step to evaluate the model you finetuned compared to the base model.
+
+## Finetuning Falcon with SkyPilot
+Finetuning Falcon-7B and Falcon-40B require GPUs with 80GB memory, but Falcon-7b-sharded requires only 40GB memory. In other words, if your GPU does not have over 40 GB memory, i.e. `accelerators: A100:1`, we suggest using `ybelkada/falcon-7b-sharded-bf16`. `tiiuae/falcon-7b` and `tiiuae/falcon-40b` will need 80 GB memory. In our examples, we use A100-80GB and A100. 
+
+Try `sky show-gpus --all` for supported GPUs.
+
+We can start the finetuning of Falcon model on Open Assistant's [Guanaco](https://huggingface.co/datasets/timdettmers/openassistant-guanaco) data **with a single command**. It will automatically find the available cheapest VM on any cloud.
+
+**To finetune using different data**, simply replace the path in `timdettmers/openassistant-guanaco`
```
Comment (romilbhardwaj) [inline llm/falcon/README.md:25]:
```suggestion
**To finetune using different data**, simply replace the path in `timdettmers/openassistant-guanaco` with any other huggingface dataset.
```

**Review by romilbhardwaj — COMMENTED** [review]
Awesome work @xzrderek! Left some comments. Will follow up offline.

**Comment by Michaelvll** [comment]
Note: one thing to look at in the future https://github.com/AdrianBZG/LLM-distributed-finetune

File: llm/falcon/README.md (hunk 1 above)
Comment (xzrderek) [reply llm/falcon/README.md:19]: I believe in the loss graph I sent you via Slack (below), I had different hyperparameters such as `max_seq_len` being a lot smaller. There are different sets of hyperparameters that work better for the different sized models, what should the default be in the YAML if that's the case?

image

Comment (romilbhardwaj) [reply llm/falcon/README.md:19]: Hmm, we should try to include the hyperparameters if we know them. How about having something like this the `run` of the YAML:

if [ $MODEL_NAME == "tiiuae/falcon-7b" ]; then
    python train.py \
    --model_name $MODEL_NAME \
    --max_seq_len 4096 \
    --bf16 \
    --group_by_length \
    --bnb_4bit_compute_dtype bfloat16 \
    --max_steps 200 \
    --dataset_name timdettmers/openassistant-guanaco
  elif [ $MODEL_NAME == "tiiuae/falcon-40b" ]; then
    python train.py \
    --model_name $MODEL_NAME \
    --max_seq_len 2048 \
    --bf16 \
    --group_by_length \
    --bnb_4bit_compute_dtype bfloat16 \
    --max_steps 200 \
    --dataset_name timdettmers/openassistant-guanaco
  fi

File: llm/falcon/train.yaml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  conda activate falcon
+  wandb login $WANDB_API_KEY
+  echo "Starting training..."
+  python train.py \
+  --model_name $MODEL_NAME \
+  --max_seq_len 2048 \
+  --bf16 \
+  --group_by_length \
+  --bnb_4bit_compute_dtype bfloat16 \
+  --max_steps 10 \
```
Comment (romilbhardwaj) [inline llm/falcon/train.yaml:40]: Should this be 200 or more? `10` seems to be too short.

File: llm/falcon/train.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    )
+
+    tokenizer = AutoTokenizer.from_pretrained(script_args.model_name, trust_remote_code=True)
+    tokenizer.pad_token = tokenizer.eos_token
+
+    return model, peft_config, tokenizer
+
+
+training_arguments = TrainingArguments(
+    output_dir=script_args.output_dir,
```
Comment (romilbhardwaj) [inline llm/falcon/train.py:183]: Quick question - does this also load checkpoints if they already exist in output_dir? (i.e., if the spot instance fails and SkyPilot restarts the job, will it continue training from the last checkpoint?)

File: llm/falcon/README.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+Currently, such `A100-80GB:1` spot instances are only available on AWS and GCP.
+
+[Optional] **To use on-demand `A100-80GB:1` instances**, which are currently available on Lambda Cloud, Azure, and GCP:
+```bash
+sky launch -c falcon -s train.yaml --no-use-spot
+```
+
+For reference, below is a loss graph you may expect to see, and the amount of time and the approximate cost of fine-tuning each of the models (assuming a spot instance GPU rate at $0.39 / hour):
+
+<img width="524" alt="image" src="https://github.com/xzrderek/skypilot/assets/32891260/cdd81781-f5b8-462b-8190-0c1da55f0526">
```
Comment (romilbhardwaj) [inline llm/falcon/README.md:50]: Should we host this on imgur.com so its more persistent? : )

File: llm/falcon/README.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+1. `ybelkada/falcon-7b-sharded-bf16`: 2.5 to 3 hours using 1 A100 GPU; total cost ≈ $1.
```
Comment (romilbhardwaj) [inline llm/falcon/README.md:52]: Q. how do we decide how long to train (2.5-3 hours)? Is this empirical? 

Another option is to list the hourly cost, and then say typical fine-tuning takes 3-4 hours or so.

File: llm/falcon/README.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+2. `tiiuae/falcon-7b`: 2.5 to 3 hours using 1 A100 GPU; total cost ≈ $1.
+
+3. `tiiuae/falcon-40b`: 10 hours using 1 A100-80GB; total cost ≈ $4.
```
Comment (romilbhardwaj) [inline llm/falcon/README.md:56]: Curious how we got $4 here. Don't A100-80GB spot instances cost  ~$1.1/hr, so for 10hrs it would be ~$11?

Comment (romilbhardwaj) [reply llm/falcon/README.md:56]: Same for other cost numbers above

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @xzrderek! Doing a run now. Left some comments.

File: llm/falcon/train.yaml (hunk 4 above)
Comment (xzrderek) [reply llm/falcon/train.yaml:40]: Ah... sorry about that, I changed it to 10 to test the checkpoint saving capabilities and forgot to change it back to 500, will fix this now.

File: llm/falcon/README.md (hunk 8 above)
Comment (xzrderek) [reply llm/falcon/README.md:56]: Yeah, this is something I was a bit confused about actually. When I start the spot job using `sky spot launch -n falcon train.yaml`, and `train.yaml` uses `accelerators: A100-80GB:1`, it gives the image below which led me to believe that a spot instance using A100-80GB costs $0.40 / hour, which is where the number comes from, is that incorrect?

File: llm/falcon/README.md (hunk 7 above)
Comment (xzrderek) [reply llm/falcon/README.md:52]: The 2.5 - 3 hours is from training 500 epochs, I'll add that information since it's not clear. I got that number from testing all three models manually.

File: llm/falcon/README.md (hunk 8 above)
Comment (romilbhardwaj) [reply llm/falcon/README.md:56]: Ah, that's the cost for the spot controller which manages the job, not the job itself.. we should fix this cost estimate. You can check the cost by running `sky show-gpus A100-80GB:1` and looking at the spot price column.

**Review by romilbhardwaj — APPROVED** [review]
Thanks @xzrderek! I pushed some changes to the readme, should be good to go now. Ran a training again, and it seems to work nicely:

image

Merging may be blocked on #2536.

**Comment by Michaelvll** [comment]
This is awesome! We can have a serving yaml for falcon as well, especially the latest falcon-180B, so people can play with the largest model open-source LLM themselves. ; )
